### PR TITLE
Rewrite of supported_os module

### PR DESCRIPTION
**Pull request by tuxudo**
Adds:
- Shipping OS
- Client tab
- Admin page with button to process all machines
- Yaml listing and widget
- Ability to update current OS without having to push out new module by having module pull and cache a JSON from module's GitHub repo

### CODE REVIEW

**Comment by tuxudo** [comment]
Should also resolve feature request #4

**Comment by bochoven** [comment]
Do you still need the json?

File: supported_os_controller.php [hunk 1]
```diff
@@ -52,5 +56,217 @@ public function os()
...
+     * Force data pull from supported_os GitHub
+     *
+     * @return void
+     * @author tuxudo
+     **/
+    public function update_cached_data()
+    {
+        // Authenticate
+        $obj = new View();
+        if (! $this->authorized()) {
```
Comment (bochoven) [inline supported_os_controller.php]: Please remove the authorised() block - not necessary anymore

File: supported_os_controller.php [hunk 2]
```diff
@@ -52,5 +56,217 @@ public function os()
...
+     * Pull in supported os data for all serial numbers :D
+     *
+     * @return void
+     * @author tuxudo
+     **/
+    public function pull_all_supported_os_data($incoming_serial = '')
+    {
+        $obj = new View();
+        // Authenticate
+        if (! $this->authorized()) {
```
Comment (bochoven) [inline supported_os_controller.php]: Please remove the authorised() block - not necessary anymore

File: supported_os_controller.php [hunk 3]
```diff
@@ -52,5 +56,217 @@ public function os()
...
+    /**
+     * Reprocess serial number
+     *
+     * @return void
+     * @author tuxudo
+     **/
+    public function recheck_highest_os($serial)
+    {
+        // Authenticate
+        if (! $this->authorized()) {
```
Comment (bochoven) [inline supported_os_controller.php]: same here

File: supported_os_controller.php [hunk 4]
```diff
@@ -52,5 +56,217 @@ public function os()
...
+    /**
+     * Return JSON with information for admin page
+     *
+     * @return void
+     * @author tuxudo
+     **/
+    public function get_admin_data()
+    {
+        $obj = new View();
+        if (! $this->authorized()) {
```
Comment (bochoven) [inline supported_os_controller.php]: etc

File: supported_os_controller.php [hunk 5]
```diff
@@ -52,5 +56,217 @@ public function os()
...
+    {
+        $obj = new View();
+
+        if (! $this->authorized()) {
+            $obj->view('json', array('msg' => 'Not authorized'));
+            return;
+        }
+
+        $os = new Supported_os_model($serial_number);
+        $obj->view('json', array('msg' => $os->rs));
```
Comment (bochoven) [inline supported_os_controller.php]: There's now jsonView() that makes this more simple: `jsonView($os->rs);`

**Comment by bochoven** [comment]
When clicking on 'process supported os' I get ```Unable to fetch new YAML from supported_os GitHub page!! Using local version instead.
Warning: Cannot modify header information - headers already sent by (output started at /Volumes/Development/thirdpartymodules/supported_os/supported_os_model.php:62) in /Volumes/Development/munkireport-php/app/helpers/site_helper.php on line 235```

File: supported_os_controller.php (hunk 1 above)
Comment (tuxudo) [reply supported_os_controller.php]: Removed

File: supported_os_controller.php (hunk 2 above)
Comment (tuxudo) [reply supported_os_controller.php]: Removed

File: supported_os_controller.php (hunk 5 above)
Comment (tuxudo) [reply supported_os_controller.php]: Changed

File: supported_os_controller.php (hunk 4 above)
Comment (tuxudo) [reply supported_os_controller.php]: Removed

File: supported_os_controller.php (hunk 2 above)
Comment (tuxudo) [reply supported_os_controller.php]: Removed

**Comment by tuxudo** [comment]
You'll get that error because the JSON file doesn't exist in the root of the master branch of the supported_os module. When it is merged, it'll be resolved. To test it, you can change the URL on line 67 of the controller and 57 of the model to `https://raw.githubusercontent.com/munkireport/supported_os/master/supported_os_data.json`

The module is still using the JSON file. I do not know how to make it work with YAML, but have changed all references of JSON to YAML within it

**Comment by bochoven** [comment]
This is how you can use the YAML parser:

```php
        $yaml_result = curl_exec($ch);

        // Check if we got results
        if (strpos($yaml_result, 'current_os') === false ){
            $yaml_data = (object) Symfony\Component\Yaml\Yaml::parseFile(__DIR__ . '/supported_os_data.yml');
            $return_status = 2;
            $cache_source = 2;
        } else {
            $yaml_data = (object) Symfony\Component\Yaml\Yaml::parse($yaml_result);
            $return_status = 1;
            $cache_source = 1;
        }
        
        $current_os = $yaml_data->current_os;
```

**Comment by tuxudo** [comment]
Module now uses YAML based cache file. All JSON references removed.

**Pull request merged**